### DISCUSSION TITLE
Storefront API Caching Issue - Stale Data for Metaobjects

### DISCUSSION THREAD

**Discussion by gulam-quinn** [Shopify/storefront-api-feedback#199] [Help]
The storefront API is caching metaobjects data for extended periods, sometimes exceeding several hours or even more than a day. Despite manually updating data via the store admin portal, the API continues to serve stale information. Any suggestions to resolve this caching problem?

This is my query
```
query MyQuery($handle: MetaobjectHandleInput){
    metaobject(handle:$handle){
        id
        fields{
            key
            value
            type
        }
        handle
        type
    }
}
```

```javascript
async function storeFrontRequest({ query, variables }) {
  const url = `https://${window.shop_domain}/api/2023-04/graphql.json`;
  const response: any = await fetch(url, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Accept: "application/json",
      "X-Shopify-Storefront-Access-Token": storefrontAccessToken,
    },
    body: JSON.stringify({
      query: query,
      variables: variables,
    }),
  });
}
```

We are facing this consistently since Shopify made storefront api available for non-channel apps. Dev change log [link](https://shopify.dev/changelog/storefront-api-is-now-available-to-non-channel-apps)

**Comment by ravirojha** [comment]
This is a genuine issue. Did you find a workaround to this? I've been facing issues due to this.

**Comment by arvindquinn** [comment]
Same here, facing this issue as well. Kind of strange that it only happens for some stores. Also, we haven't seen this same issue for metafields, only happens for metaobjects.

**Comment by johncraigcole** [comment]
Hey Everyone 👋 . For this type of investigation it would be best to contact support so we can make sure your authenticated on the store in question and be able to discuss details about the store. If this is reproducible, including the following in the support request would be helpful.

1. The date and time in UTC the metaobject was updated  along with id of the metaobject.
2. The `x-request-id` from the response headers of the Storefront API request where stale data was returned.

We have a limited log retention window so sharing the above with support on the day the issue is reproduced would be beneficial. If you already have a ticket with support, please share the ticket number and I will try and help out here.

**Comment by arvindquinn** [comment]
Thanks for getting back on this.

We haven't encountered this issue again this entire week so I think it's
solved somehow with no changes on our end.

In case we have an issue like this in the future, will share those extra
data you've asked for as well to make debugging easier.

Thanks for all your help!

Regards
Arvind

**Comment by albyonl** [comment]
Still facing this issue

**Reply by Enkratia** [reply]
Hello, did you find any solution?

**Reply by albyonl** [reply]
there isn't a solution for it - just need to wait for shopify to fix

**Reply by Enkratia** [reply]
@albyonl actually this is working as expected, because Next.js uses "force-cache" by default. To change this behaviour, you just need to add these lines:
```
      next: {
        revalidate: 60, // 60 seconds
      },
```

context (where to put these lines):
```
    const response = await fetch(args, {
      next: {
        revalidate: 60,
      },
      headers: {
        "Content-Type": "application/json",
        "X-Shopify-Storefront-Access-Token": storefront_key,
        "X-Shopify-Access-Token": admin_key,
      },
      method: "POST",
      body,
    });
```

**Reply by albyonl** [reply]
this is a different issue. shopify uses its own cache on edge servers, which causes the problem discussed in this thread. you can confirm this by checking the headers returned from the Shopify request.

**Reply by gulam-quinn** [reply]
There is nothing wrong with how you're fetching the data; this is definitely an issue on Shopify's side, as it was in our case. The problem was caused by a bug in Shopify’s cache invalidation, which led the Storefront API to return cached data in certain data center regions. You can check the specific data center by inspecting the `X-DC` header in the Storefront API response, and you may need to share this region value with Partner Support for quicker resolution. As @johncraigcole suggested, reaching out to Partner Support is recommended for immediate help.

**Reply by adrianxadamn** [reply]
@albyonl @gulam-quinn have you gotten any support for this? Have you reached out to Partner Support?

**Reply by gulam-quinn** [reply]
I honestly don’t remember if we ever got any update from the support team, it was more than two years ago. But we never faced this issue again after that

**Reply by adrianxadamn** [reply]
@gulam-quinn  I realized it was an implementation issue on my part. Thanks for the response
